webapp/src/ts/services/cht-script-api.service.ts

```diff
@@ -0,0 +1,114 @@
+import { Injectable, OnDestroy } from '@angular/core';
+import { Subscription } from 'rxjs';
+
+import { UserSettingsService } from '@mm-services/user-settings.service';
+import { SettingsService } from '@mm-services/settings.service';
+import { ChangesService } from '@mm-services/changes.service';
+import { SessionService } from '@mm-services/session.service';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class CHTScriptApiService implements OnDestroy {
+  private userSettingsDoc;
+  private settings;
+  subscriptions: Subscription = new Subscription();
+
+  constructor(
+    private userSettingsService: UserSettingsService,
+    private settingsService: SettingsService,
+    private sessionService : SessionService,
+    private changesService: ChangesService
+  ) { }
+
+  init() {
+    this.watchChanges();
+
+    return this
+      .getSettings()
+      .then(() => this.getUserSettingsDoc());
+  }
+
+  ngOnDestroy(): void {
+    this.subscriptions.unsubscribe();
+  }
+
+  private getUserSettingsDoc() {
+    return this.userSettingsService
+      .get()
+      .then(userSettingsDoc => {
+        this.userSettingsDoc = userSettingsDoc;
+      });
+  }
+
+  private getSettings() {
+    return this.settingsService
+      .get()
+      .then(settings => {
+        this.settings = settings;
+      });
+  }
+
+  private watchChanges() {
+    const userDocId = this.userSettingsService.getUserDocId();
+
+    const userSettingsSubscription = this.changesService.subscribe({
+      key: 'cht-script-api-user-settings-changes',
+      filter: change => change.id === userDocId,
+      callback: () => this.getUserSettingsDoc()
+    });
+    this.subscriptions.add(userSettingsSubscription);
+
+    const settingsSubscription = this.changesService.subscribe({
+      key: 'cht-script-api-settings-changes',
+      filter: change => change.id === 'settings',
+      callback: () => this.getSettings()
+    });
+    this.subscriptions.add(settingsSubscription);
+  }
+
+  private hasRole(role: string): boolean {
+    if (!this.userSettingsDoc?.roles?.length) {
+      return false;
+    }
+
+    return this.userSettingsDoc.roles.includes(role);
+  }
+
+  private hasPermission(permission: string): boolean {
+    if (!this.userSettingsDoc?.roles?.length) {
+      return false;
+    }
+
+    if (this.sessionService.isAdmin()) {
+      // Admin has the permissions automatically.
+      return true;
+    }
+
+    const roles = this.settings.permissions[permission];
+
+    if (!roles) {
+      return false;
+    }
+
+    return this.userSettingsDoc.roles.some(role => roles.includes(role));
+  }
```

Comment:
This code appears to be copied from the Auth service or similar. We need to reuse this rather than copying it because it's easy to get it wrong. This could be as simple as one of these services calling the other, or this could be extracted into a shared lib which might then also reduce the duplication in api, admin, sentinel, etc.

Reply:
Yeah, it's a simplified sync version of the one in Auth service, without checking for disallowed permissions (as Auth service do) which I haven't found usages of it in webapp  🤔

I thought about that and actually tried to use the one in Auth service but it needs to be simplified and refactored so we have the async function and sync function that uses the same checking function 😁 If that okay I'll go and refactor Auth Service code.

Reply:
Done